lib/proxy/router.go

```diff
@@ -313,7 +314,32 @@ func (r *Router) DialHost(ctx context.Context, clientSrcAddr, clientDstAddr net.
 		return nil, trace.Wrap(err)
 	}
 
-	return NewProxiedMetricConn(conn), trace.Wrap(err)
+	return NewProxiedMetricConn(&checkedPrefixWriter{
+		Conn: conn,
+		// SSH connection MUST start with "SSH-2.0" bytes according to https://datatracker.ietf.org/doc/html/rfc4253#section-4.2
+		requiredPrefix: []byte("SSH-2.0"),
+	}), trace.Wrap(err)
+}
+
+// checkedPrefixWriter checks that first data written into it has the specified prefix.
+type checkedPrefixWriter struct {
+	net.Conn
+
+	requiredPrefix []byte
+	checked        bool
+}
+
+func (c *checkedPrefixWriter) Write(p []byte) (n int, err error) {
+	if c.checked {
+		return c.Conn.Write(p)
+	}
+
+	// It is assumed that required prefix is small enough to always fit into data provided on the first write.
```

Comment:
This is kind of a bad assumption and it will lead to really strange issues precisely in situations where connectivity is already bad enough to be annoying in other ways.

I recommend changing `requiredPrefix` and `checked` into `requiredData`, and then you can check if the beginning of `p` matches the beginning of `requiredData` (depending on which one is the shorter), then writing, then advancing `requiredData` _by the returned amount by Write_.

Right now you can bypass this (accidentally or not):

```go
w := &checkedPrefixWriter{requiredPrefix: p}

w.SetWriteDeadline(time.UnixNano(1))
w.Write(p) // guaranteed to fail, but sets "checked" to true
w.SetWriteDeadline(time.Time{})
w.Write(notP)
```